### PR TITLE
test: a test framework for source parsing

### PR DESCRIPTION
**Pull request by neverchanje**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

A framework for testing source parsing. It will generate all possible types of data and verify that the parsed record are equal to the generated record. This PR only supports generating data in JSON format.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### CODE REVIEW

File: src/source/parser_test/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        .into(),
+        T::Time => NaiveTimeWrapper(NaiveTime::from_hms(
+            rng.gen_range(0..=23),
+            rng.gen_range(0..=59),
+            rng.gen_range(0..=59),
+        ))
+        .into(),
+        T::Timestamp | T::Timestampz => NaiveDateTimeWrapper(NaiveDateTime::from_timestamp(
+            rng.gen_range(1000000000..=1662346644),
+            rng.gen_range(0..1_000_000_000),
```
Comment (jon-chuang) [inline src/source/parser_test/src/lib.rs:98]: we actually want to round up nanoseconds to microseconds, as we only support up to microseconds now.

Comment (jon-chuang) [reply src/source/parser_test/src/lib.rs:98]: Can use e.g.: https://docs.rs/chrono/latest/chrono/naive/struct.NaiveDateTime.html#method.duration_round

**Comment by jon-chuang** [comment]
btw, this seems to duplicate alot of code with connector/source/datagen. This is based on generation logic in common/src/field_generator

**Comment by neverchanje** [comment]
@jon-chuang Thanks! I will take a look and see if I can reuse the code in field_generator.

File: src/source/parser_test/src/lib.rs (hunk 1 above)
Comment (jon-chuang) [reply src/source/parser_test/src/lib.rs:98]: btw, if you use the field_generator, I alr do the rounding here: https://github.com/risingwavelabs/risingwave/pull/5514

**Comment by neverchanje** [comment]
@jon-chuang Since `field_generator` doesn't cover all the data types, I cannot directly reuse the code. I feel like the refactoring can be a to-do and the implementation in this PR is not very complex.

**Comment by jon-chuang** [comment]
Hmm, my worry is that it simply duplicates a lot of logic, and a refactor is more of a rewrite. 

Hence, I think it may be a good idea if you could explore implementing the missing features in FieldGenerator, and then build on top of what you've done in this PR based on FieldGenerator. What do you think?

**Comment by neverchanje** [comment]
> Actually, I don't completely understand the purpose of this framework. What tests specifically do we want to run? Is it more of a fuzz-test? Why is the current testing inadequate?

Yes. It's fuzzy testing. I want to use this framework to ensure all possible data input can be well handled by the source parser. Indeed, it currently can't cover some corner cases. At the time I wrote this PR, all the test cases are normal cases, so I thought at least the new test can cover more. But I just noticed that you added some corner cases. I will revert the original test back.

> Finally, I do not see that this PR implements fault injection and checking that those faulty data are dropped. How should we proceed with that? What kind of faults are we injecting?

It's not the purpose of this PR. @wangrunji0408 is working on it. He will replace rdkafka with its madsim-backed fake rdkafka when running deterministic tests. #5181

**Comment by jon-chuang** [comment]
> I want to use this framework to ensure all possible data input can be well handled by the source parser.

I agree on this goal, but do we expect this framework to be able to help us identify previously unknown problematic inputs? Take the float without decimals example for instance. In this case, it's hard to imagine this framework would have identified the issue.

I can understand this form of testing as a second line of defense after unit testing. But could you provide examples in which you believe totally random inputs could help us find bugs? Perhaps there is a framework that can help us provide corner cases rather than just totally random inputs? 

I would also like to suggest exploring usage of frameworks like proptest which may be able to find minimal failing examples if it finds a bug https://altsysrq.github.io/proptest-book/proptest/getting-started.html

**Pull request closed without merging**